### ISSUE TITLE
Support deallocate mode for Azure vmss when shutting down

### ISSUE THREAD

**Issue by nclaeys** [kubernetes/autoscaler#5923]
**Which component are you using?**:

Cluster autoscaler for Azure

**Is your feature request designed to solve a problem? If so describe the problem this feature should solve.**:

Azure does not support customizing the node image, but it does allow to stop (Deallocate) nodes in a nodepool. This would drastically speed up the startup of our nodes using sysbox (which is now installed by a deamonset at startup).  https://github.com/nestybox/sysbox/blob/master/docs/user-guide/install-k8s.md

**Describe the solution you'd like.**:

- We should be able to configure a given vmss with a tag/label to indicate it should Deallocate a node instead of Delete it (delete should be the default).
- In the scale_down code we should use this property to either call DeleteInstances or DeallocateInstances for a vmss.
- in the scale_up: first start deallocated nodes before scaling up the vmss

**Describe any alternative solutions you've considered.**:

- First looked at customizing the node image (this was our preferred option), but this is not allowad by Azure, only the capi project  or the legacy aks-engine project support it.

**Additional context.**:
In the Azure docs about the shutdownMode, they mention it as their solution to speed up the initial node creation. https://learn.microsoft.com/en-us/azure/aks/scale-down-mode

**Comment by Bryce-Soghigian** [comment]
We already support deallocated mode in the AKS Cluster autoscaler fork. As you linked, one can configure the scale-down-mode in the Agentpool api, to configure their nodepool to use deallocate mode. 

Ideally, we would like to bring these changes to the upstream, but there is a lot involved in our deallocate mode feature, and its specific to our cloud provider. I am thinking it might be better to implement it at the cloud provider level, whereas now its built into scale down. 




# Scale down deallocate mode logic inside the aks-fork 
<img width="289" alt="Screenshot 2023-07-04 at 12 48 31 PM" src="https://github.com/kubernetes/autoscaler/assets/49734722/292c2789-6826-4530-b038-2d0f7ec035b6">

<img width="488" alt="Screenshot 2023-07-04 at 12 48 14 PM" src="https://github.com/kubernetes/autoscaler/assets/49734722/4324fb7e-ed80-49fb-9cef-f2d9564119bd">

**Comment by nclaeys** [comment]
Hi Bryce,

From the Azure docs I seemed indeed that it should be working. Is this fork open source, such that I can take a look?

We run an unmanaged cluster autoscaler because of some issues with the manage one:
- need to support scale down to 0 for all our nodepools: https://github.com/Azure/AKS/issues/2976
- support all configuration options for the cluster autoscaler
- allows us to easily migrate to nodepools when upgrading to newer kubernetes version

**Comment by Bryce-Soghigian** [comment]
Yes, it only works with the aks-fork of the autoscaler in our managed autoscaler. Can you share your configuration options you set, and maybe if they are shared with other customers i can bring them back to the managed AKS fork, and nodepool api? No promises, just if there is enough overlap with other customers. 

We support scale from zero with our managed autoscaler as well.  And we have auto upgrader for easy migration of nodepools on new k8s versions, that i have found useful for my own personal clusters upgrade. 

We do not support scale from zero or deallocate mode upstream, but both are supported in our fork. I would like to move things out of our fork, but in a way where upstream will accept these changes, and it won't be too disruptive for other cloud providers. 

Could you outline 
1. Flags you use in your autoscaler configuration 
2. Where are the migration issue pain points when using auto upgrader, I can see if I can relay them to the upgrade sig inside AKS.

**Comment by nclaeys** [comment]
Hi Bryce,

Most of our effort related to the cluster-autoscaler on AKS dates from last year. Back then we had too many issues at our customers due to the taints causing scale up to fail. ALso back then the managed cluster-autoscaler did not allow most of the configuration options (e.g. scale-down-unneeded, scale-down-utilization-threshold,...). As far as I know, at the moment you support all necessary configuration options.

Our main issue still remains the fact that we cannot specify user-defined taints to ignore (apart from the remediator taint), which impacts the cluster-autoscaler scaling up from 0. We have our own taints, which we can change to include the ignoreTaintsPrefix, but also use for example sysbox on certain nodepools which has its own taints. For these it is not feasible to fork the project just to rename the taint.

We indeed do not use the auto upgrader provided, mainly because of 2 issues which has to do to minimize the potential impact on our customers. 
1. Certain of the components we deploy at customers are tight to specific k8s version, which is why we want to keep control of upgrading those after testing them out on our internal clusters before rolling it out at our customers.
2. Maybe even more importantly is to reduce the impact of the rollout at our customers. Since we run primarily batch jobs at our customers, we cannot just drain all the nodes, as then these jobs would fail which is not what we want. To me it is not completely clear how the auto-upgrade works internally which is why I do not want to rely on it. We want:
  - cordon old nodes
  - create new nodes
  - delete the old nodes only when they are unneeded

A last suggestion or question is that to me it is not at all clear on what improvements Azure makes with regard to the internal components. For example I always assumed you are running a fork of the cluster-autoscaler but where the functionality differs from the upstream repository is not at all clear to me. I find it a pity that you use this open-source component, which has been developed by the community, forked it and do not make it available to the community.

**Comment by Bryce-Soghigian** [comment]
Thanks for providing this perspective. We are committed to lowering the difference as much as possible between our managed autoscaler and the upstream autoscaler, but it's a gradual effort, and ongoing with each new Kubernetes version.

**Comment by nclaeys** [comment]
@jackfrancis Sorry for the late response. From a user perspective this looks great. This implements the functionality that I would like to see when using vmss. Thanks for contributing this :+1: !

**Comment by jackfrancis** [comment]
@nclaeys I'm planning to demo this and propose to the community next Monday Jan 22 at 16:00 Poland time zone (details [here](https://github.com/kubernetes/community/blob/master/sig-autoscaling/README.md#meetings)) if you'd like to attend and add feedback

**Comment by jackfrancis** [comment]
@nclaeys full disclosure this feature has been deprioritized and probably won't get picked back up

Are you unable to use AKS + Autoscaler for this need?

cc @comtalyst

**Comment by nclaeys** [comment]
That is a pity, was hoping this would get in there. I will have to test out whether the AKS autoscaler will support all the configuration changes that we have made.
Based on a quick investigation:
- priority expander should work
- upgrades we could handle by adding nodepools and disable that the old nodepools are used by the autoscaler
- logging tweaks seem not to be possible with the AKS managed autoscaler, correct?
- startup taints seem not to be exposed by the aks managed autoscaler, this will make it currently impossible to use for us.

Is there a plan to make the AKS managed autoscaler more configurable?
